driver-core/src/test/unit/com/mongodb/internal/connection/InternalStreamConnectionInitializerSpecification.groovy

```diff
@@ -269,4 +427,65 @@ class InternalStreamConnectionInitializerSpecification extends Specification {
+        def finalServerResponse = 'v=rmF9pqV8S7suAoZWja4dJRkFsKQ='
+        enqueueSpeculativeAuthenticationResponsesForScramSha(initialServerResponse, finalServerResponse)
+    }
+
+    def enqueueSpeculativeAuthenticationResponsesForScramSha(final String initialServerResponse,
+                                                             final String finalServerResponse) {
+        internalConnection.enqueueReply(buildSuccessfulReply(
+                '{ok: 1, maxWireVersion: 9, ' +
+                        'ismaster: true, ' +
+                        'speculativeAuthenticate: { conversationId: 1, done: false, ' +
+                        "payload: BinData(0, '${encode64(initialServerResponse)}')}}"))
+        internalConnection.enqueueReply(buildSuccessfulReply(
+                '{ok: 1, maxWireVersion: 9, ' +
+                        'conversationId: 1, done: true, ' +
+                        "payload: BinData(0, '${encode64(finalServerResponse)}')}"))
+        internalConnection.enqueueReply(buildSuccessfulReply('{ok: 1}'))
+    }
+
+    def enqueueSpeculativeAuthenticationResponsesForX509() {
+        internalConnection.enqueueReply(buildSuccessfulReply(
+                '{ok: 1, maxWireVersion: 9, ismaster: true, conversationId: 1, ' +
+                        'speculativeAuthenticate: { dbname: \"$external\", ' +
+                        'user: \"CN=client,OU=KernelUser,O=MongoDB,L=New York City,ST=New York,C=US\" }}'))
+        internalConnection.enqueueReply(buildSuccessfulReply('{ok: 1}'))
+    }
+
+    def enqueueSpeculativeAuthenticationResponsesForPlain() {
+        internalConnection.enqueueReply(buildSuccessfulReply(
+                '{ok: 1, maxWireVersion: 9, ismaster: true, conversationId: 1}'))
+        internalConnection.enqueueReply(buildSuccessfulReply(
+                '{ok: 1, done: true, conversationId: 1}'))
+        internalConnection.enqueueReply(buildSuccessfulReply('{ok: 1}'))
+    }
+
+    def encode64(String string) {
+        Base64.encode(string.getBytes(Charset.forName('UTF-8')))
+    }
+
+    def validateIsMasterCommand(final String firstClientChallenge, final String mechanism,
+                                final boolean hasSaslSupportedMechs) {
+        String isMaster = '{ismaster: 1, ' +
+                (hasSaslSupportedMechs ? 'saslSupportedMechs: "database.user", ' : '') +
+                (mechanism == 'MONGODB-X509' ?
+                        'speculativeAuthenticate: { authenticate: 1, ' +
+                                "mechanism: '${mechanism}', db: \"\$external\" } }" :
+                        'speculativeAuthenticate: { saslStart: 1, ' +
+                                "mechanism: '${mechanism}', payload: BinData(0, '${encode64(firstClientChallenge)}'), " +
+                                'db: "admin", options: { skipEmptyExchange: true } } }')
+
+        assertEquals(BsonDocument.parse(isMaster), decodeCommand(internalConnection.getSent()[0]))
```

Comment:
I'm not sure that this works in Spock, to make an assertion in a method like this.  And even if it does, it's a bit obfuscated to do so.

Consider changing this method to just return the result of `BsonDocument.parse(isMaster)`, and do the actual assertion in the then clause using `==`